services/agent/src/simcore_service_agent/services/containers_manager.py

```diff
@@ -0,0 +1,83 @@
+import logging
+from dataclasses import dataclass, field
+
+from aiodocker import Docker
+from fastapi import FastAPI
+from models_library.api_schemas_directorv2.services import (
+    DYNAMIC_PROXY_SERVICE_PREFIX,
+    DYNAMIC_SIDECAR_SERVICE_PREFIX,
+)
+from models_library.projects_nodes_io import NodeID
+from servicelib.fastapi.app_state import SingletonInAppStateMixin
+from servicelib.utils import limited_gather
+
+from ..core.settings import ApplicationSettings
+from .docker_utils import get_containers_with_prefixes, remove_container_forcefully
+
+_logger = logging.getLogger(__name__)
+_always_visible_logger = logging.getLogger(__name__)
+
+
+@dataclass
+class ContainersManager(SingletonInAppStateMixin):
+    app_state_name: str = "containers_manager"
+
+    docker: Docker = field(default_factory=Docker)
+
+    async def force_container_cleanup(self, node_id: NodeID) -> None:
+        # compose all possible used container prefixes
+        proxy_prefix = f"{DYNAMIC_PROXY_SERVICE_PREFIX}_{node_id}"
+        dy_sidecar_prefix = f"{DYNAMIC_SIDECAR_SERVICE_PREFIX}_{node_id}"
+        user_service_prefix = f"{DYNAMIC_SIDECAR_SERVICE_PREFIX}-{node_id}"
```

Comment:
Q: the difference between the value of `user_service_prefix` and `dy_sidecar_prefix` is `-` vs. `_`, why? Why is the name of one "user" and the other "dy-sidecar"?

Reply:
So this all comes down to the naming conventions we use to distinguish various containers we are running.

- `proxy_prefix` is the naming convention we use for the proxy that is started together with the dynamic-sidecar, named something like: `dy-proxy_UUID`
- `dy_sidecar_prefix` is the dynamic-sidecar named like: `dy-sidecar_UUID`
- `user_service_prefix` will detected all the user services (notice the plural) started by the dynamic-sidecar. By convention and to make our life simpler during debugging, the containers are named with the following pattern `dy-sidecar-UUID.NUMBER.RANDOM_DOCKER_STRING` where we do not have any say in the part that comes after `dy-sidecar-UUID`. You have correctly pointed out the difference which is a `-` instead of `_`. Notice: services like s4l might have up to 4 containers here. Advantages of doing this:
  - we can easily identify which dynamic-sidecar started the user service
  - we can easily find all the logs of `sidecar + user services`: by filtering via the `f"UUID"`
  - we can easily obtain only the logs of the user services by filtering via `UUID-`